### ISSUE TITLE
path_complete doesn't work with CommandSet

### ISSUE THREAD

**Issue by muhrin** [python-cmd2/cmd2#967]
Version 1.3.0.

When using `path_complete` for a `CommandSet` `argparse` argument the following exception occurs:

```
Traceback (most recent call last):
  File "/home/martin/.virtualenvs/pyos/lib/python3.8/site-packages/cmd2/cmd2.py", line 1707, in complete
    self._completion_for_command(text, line, begidx, endidx, shortcut_to_restore)
  File "/home/martin/.virtualenvs/pyos/lib/python3.8/site-packages/cmd2/cmd2.py", line 1592, in _completion_for_command
    self.completion_matches = self._redirect_complete(text, line, begidx, endidx, compfunc)
  File "/home/martin/.virtualenvs/pyos/lib/python3.8/site-packages/cmd2/cmd2.py", line 1385, in _redirect_complete
    return compfunc(text, line, begidx, endidx)
  File "/home/martin/.virtualenvs/pyos/lib/python3.8/site-packages/cmd2/cmd2.py", line 1757, in _complete_argparse_command
    return completer.complete_command(tokens_to_parse, text, line, begidx, endidx, cmd_set=cmd_set)
  File "/home/martin/.virtualenvs/pyos/lib/python3.8/site-packages/cmd2/argparse_completer.py", line 429, in complete_command
    completion_results = self._complete_for_arg(pos_arg_state.action, text, line,
  File "/home/martin/.virtualenvs/pyos/lib/python3.8/site-packages/cmd2/argparse_completer.py", line 596, in _complete_for_arg
    results = arg_choices.to_call(*args, **kwargs)
TypeError: path_complete() takes 5 positional arguments but 6 were given
```

This happens because the `CommandSet` is being passed in as the second positional argument which isn't expected by the completer.

A big benefit of completers is that they can be used by disparate commands (whether they be free functions later bound to a `Cmd` class, members of a `Cmd` class or members of a `CommandSet`).  Having a different interface detracts from this as, now, something like a shim is required to make it work without redefining a new `path_complete`.

I think it's worth considering ways to have a common interface, not only for the complete functions but for `do` and `help`.  A couple of suggestions come to mind:

1.

`def path_complete(self, text: str, line: str, begidx: int, endidx: int, context)`

where `context` is a dict or `namedtuple` that stores a reference to the `Cmd` instance and (optionally) the `CommandSet`.

2.

Both `Cmd` and `CommandSet` conform to a common interface that specifies the subset of things that can be done by both (most methods/attributes from `Cmd`), and `CommandSet` just forwards these to its `_cmd` member.

**Comment by anselor** [comment]
@muhrin Can you provide a code snippet to reproduce this?

**Comment by muhrin** [comment]
Sure, here's an MWE:

```
#!/usr/bin/env python
"""A simple cmd2 application."""
import argparse
import cmd2


class FirstApp(cmd2.Cmd):
    pass

class MyCommands(cmd2.CommandSet):
    parser = argparse.ArgumentParser()
    parser.add_argument('path', nargs='+', help='paths', completer_method=cmd2.Cmd.path_complete)

    @cmd2.with_argparser(parser)
    def do_path(self, app: cmd2.Cmd, args):
        app.poutput(args.path)


if __name__ == '__main__':
    import sys
    c = FirstApp()
    sys.exit(c.cmdloop())
```

**Comment by anselor** [comment]
@muhrin Ah yes, it seems that the assumption I made here is that if the command is in a CommandSet then the completer_method is a member of the same CommandSet. This will require some thinking to resolve. 

Part of the complication here is the point at which completers are registered with the argparse object is when the class definition is loaded. The completers are registered as unbound functions that happen to be defined inside of a class. They don't become methods until a class instance is created later. So what our code needs to do is guess at what needs to be passed in as self so that it can behave as if it is a bound method call on an object instance. 

I'm going to experiment with a few things to see if I can make a reliable determination.

**Comment by anselor** [comment]
After some analysis of the issue and some discussion with @kmvanbrunt, I think we have a path forward. 

It really doesn't make sense to have the exact same function signature for the `do_/complete_/help_` functions in a `CommandSet` because it violates the contract of what `self` means inside of a class. The `self` parameter of a class function should be the instance of the class in which the function is defined. To match the exact signature, as you're suggesting, we would pass the cmd2 application in for `self` for the `do_/complete_/help_` functions.

For the argparse parameter completers, that is definitely an edge case that I hadn't considered. I'm adding code to try and attempt to locate the instance that matches the CommandSet type in which a completer function is defined. If it's not defined in a CommandSet, it won't be included in the arguments to the function call.

**Comment by anselor** [comment]
@muhrin 

I just pushed up changes in this branch:
https://github.com/python-cmd2/cmd2/tree/cmdset_completers

It should detect whether the completer function is in a CommandSet before passing in a CommandSet as a parameter. It'll also search all available registered CommandSets to locate the matching instance so you should be able to use a completer function across CommandSets as long as they're registered.

**Comment by muhrin** [comment]
Thanks for the quick work @anselor.

I don't think I was clear in my initial description.  I 100% agree that self should always be the class instance, anything else would be insane.  However, it's totally fine to have a different self (`Cmd` or `CommandSet`) depending on where it was defined but have the rest of the function signature be identical (which is what I was suggesting).  In fact, this is exactly what abstract methods are (i.e. a type of single dispatch based on the runtime type of the first argument - `self`, and indeed what you're implmenting is basically a single dispatch mechanism).

The solution in your branch does indeed does appear to solve the problem highlighted by this issue and in the current design it can be though of as an edge case, however I see it as part of a wider design suboptimality but I think that discussion is best left for #943 so I'll post there.

Btw, is my reading of the tests correct in that there are currently no tests for `CommandSet`?

**Comment by anselor** [comment]
@muhrin I'm not sure what you're looking at. There are a bunch of tests I wrote while implementing CommandSet. On this particular branch  it looks like it failed some coding style validation and didn't run the tests so you might not be seeing the right coverage output.

**Comment by muhrin** [comment]
Ah, sorry I think I missed the tests.  I was looking in `tests` but hasn't seen the `isolated_tests` folder.  Thanks!

**Comment by anselor** [comment]
I think since that's causing some confusion, I'm going to rename the directory to `tests_isolated` so it'll be grouped together and be more noticeable.

**Comment by anselor** [comment]
This was fixed in #968
